### DISCUSSION TITLE
Saving 1D lineout of 3D data during runtime

### DISCUSSION THREAD

**Discussion by wjwhit** [AMReX-Codes/amrex#3537] [Q&A] [answered]
Hello, 

I am wondering if there is a convenient way to extract 2D planes and/or 1D lines from 3D data during runtime. I need lineout and plane data at very high frequencies, so saving 3D data and post-processing is not feasible. On a single level, using ijk indexing is good enough for lineouts as I only need the specific location to within a grid cell, but on multiple levels it becomes more challenging. 

Thank you!

**Comment by drummerdoc** [comment, marked as answer]
In the Pele codes, there is a base class for such diagnostics that can be extended...currently provides planes, integrals, pdfs and such.  See https://github.com/AMReX-Combustion/PelePhysics/tree/91db46315b87364c45316861056a5bc47ba27c39/Utility/Diagnostics.  Let me know if you need help understanding or incorporating.  I'm sure other apps have done similar things we could use to find a nice general approach to this.

**Reply by baperry2** [reply]
The code @drummerdoc linked is a shared utility that is used by two different solvers (PeleC and PeleLMeX) and should be reasonably general for AMReX codes. You can also see the hooks to call this utility from a solver here: https://github.com/AMReX-Combustion/PeleLMeX/blob/development/Source/PeleLMDiagnostics.cpp

And the documentation such as it is can be found here: https://amrex-combustion.github.io/PeleLMeX/manual/html/LMeXControls.html#run-time-diagnostics

**Reply by wjwhit** [reply]
Thank you, both of you!

**Reply by zingale** [reply]
How hard would it be to move this functionality into AMReX itself to make it easier for other application codes to use it?

**Reply by drummerdoc** [reply]
Not "hard", just would take a little design effort, coding and testing.  We'll add it to our list todo.  We have a number of additional tools we could wire into it as well, including my AMR isosurfacing and streamline tools, so this could be a nice platform to transfer some of that over as well.
